docs/source/advanced/loops.rst

```diff
@@ -0,0 +1,399 @@
+    trainer = Trainer()
+
+    trainer.fit_loop = loop
+    # or
+    trainer.validate_loop = loop
+    # or
+    trainer.test_loop = loop
+    # or
+    trainer.predict_loop = loop
+
+The second way is when a custom loop is meant to replace one of the subloops of a top-level loop.
+In this case one can use the :meth:`~pytorch_lightning.loops.base.Loop.connect` method of a parent loop to connect/replace a child loop.
+For example:
+
+.. code-block:: python
+
+    loop = CustomLoop()
+    trainer = Trainer()
+
+    trainer.fit_loop.connect(epoch_loop=loop)
+
+To illustrate the power of loop customization we will look at a relatively simple custom loop that converts the training_step hook to a generator.
+
+Example: YieldLoop
+^^^^^^^^^^^^^^^^^^
+
+In this example, we will build a custom loop that enables us to write a new flavor of a training step, where the training step actually becomes a generator and instead of returning losses for optimization, we yield them!
+**Note:** This assumes knowledge of generators in Python and the :code:`yield` mechanism.
+
+Imagine you wanted your LightningModule training step to look like this:
+
+.. code-block:: python
+
+    def training_step(self, batch, batch_idx):
+        # do something with optimizer 0
+        loss0 = self.loss(self(batch))
+        yield loss0
+
+        # do something with optimizer 1 that requires loss0
+        loss1 = self.foo(loss0)
+        yield loss1
+
+
+Normally, we would implement a training step with a signature :code:`training_step(self, batch, batch_idx, optimizer_idx)` and then conditionally compute either :code:`loss0` or :code:`loss1` depending on the current optimizer index and return that loss at the end.
+But if the computation of say :code:`loss1` depends on :code:`loss0` or another quantity computed for the first optimizer, we would have to recompute the value for :code:`optimizer_idx = 1` and that is wasteful.
+
+With the training step as a generator as shown above however, we are able to retain the local variables across training_step boundaries when we switch from one optimizer to the next.
+The alternative to this would be *manual optimization* where the same can be achieved, but with the generator loop we can still get all benefits of manual optimization without having to call backward or zero grad ourselves.
+
+In order to enable returning a generator from a training step, we need to implement a custom loop!
+This will be a subclass of the existing :class:`~pytorch_lightning.loops.optimization.optimizer_loop.OptimizerLoop` and then be attached to the :class:`~pytorch_lightning.loops.batch.training_batch_loop.TrainingBatchLoop`.
+
+.. code-block:: python
+
+    from functools import partial
+    from pytorch_lightning.loops import Loop, OptimizerLoop
+    from pytorch_lightning.loops.optimization.optimizer_loop import ClosureResult
+    from pytorch_lightning.loops.utilities import _build_training_step_kwargs
+
+
+    class YieldLoop(OptimizerLoop):
```

Comment:
@awaelchli, would mind making sure the code isn't too long. Maybe 80 lines. It doesn't render very well in the doc.

<img width="806" alt="Screenshot 2021-09-18 at 12 18 17" src="https://user-images.githubusercontent.com/12861981/133886985-d6cefec4-31eb-479b-8826-c8dc06dafbe4.png">